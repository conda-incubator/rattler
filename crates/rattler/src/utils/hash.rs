use digest::{Digest, Output};
use sha2::Sha256;
use std::fs::File;
use std::io::Write;
use std::path::Path;

/// Compute the SHA256 hash of the file at the specified location.
pub fn compute_file_sha256(path: &Path) -> Result<sha2::digest::Output<Sha256>, std::io::Error> {
    // Open the file for reading
    let mut file = File::open(path)?;

    // Determine the hash of the file on disk
    let mut hasher = Sha256::new();
    std::io::copy(&mut file, &mut hasher)?;

    Ok(hasher.finalize())
}

/// Parses a SHA256 hex string to a digest.
pub fn parse_sha256_from_hex(str: &str) -> Option<Output<Sha256>> {
    let mut sha256 = <Output<Sha256>>::default();
    match hex::decode_to_slice(str, &mut sha256) {
        Ok(_) => Some(sha256),
        Err(_) => None,
    }
}

/// A simple object that provides a [`Write`] implementation that also immediately hashes the bytes
/// written to it.
pub struct HashingWriter<W: Write, D: Digest> {
    writer: W,
    hasher: D,
}

pub type Sha256HashingWriter<T> = HashingWriter<T, sha2::Sha256>;

impl<W: Write, D: Digest + Default> HashingWriter<W, D> {
    /// Constructs a new instance from a writer and a new (empty) hasher.
    pub fn new(writer: W) -> Self {
        Self {
            writer,
            hasher: Default::default(),
        }
    }
}

impl<W: Write, D: Digest> HashingWriter<W, D> {
    pub fn finalize(self) -> (W, Output<D>) {
        (self.writer, self.hasher.finalize())
    }
}

impl<W: Write, D: Digest> Write for HashingWriter<W, D> {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        let bytes = self.writer.write(buf)?;
        self.hasher.update(&buf[..bytes]);
        Ok(bytes)
    }

    fn flush(&mut self) -> std::io::Result<()> {
        self.writer.flush()
    }
}

#[cfg(test)]
mod test {
    use rstest::rstest;

    #[rstest]
    #[case(
        "1234567890",
        "c775e7b757ede630cd0aa1113bd102661ab38829ca52a6422ab782862f268646"
    )]
    #[case(
        "Hello, world!",
        "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3"
    )]
    fn test_compute_file_sha256(#[case] input: &str, #[case] expected_hash: &str) {
        // Write a known value to a temporary file and verify that the compute hash matches what we would
        // expect.

        let temp_dir = tempfile::tempdir().unwrap();
        let file_path = temp_dir.path().join("test");
        std::fs::write(&file_path, input).unwrap();
        let hash = super::compute_file_sha256(&file_path).unwrap();

        assert_eq!(format!("{hash:x}"), expected_hash)
    }
}
