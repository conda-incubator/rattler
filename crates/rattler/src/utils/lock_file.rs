use std::path::{Path, PathBuf};

/// Provides a cross-platform file locking implementation.
pub struct LockFile {
    path: PathBuf,
    lock: Option<fslock::LockFile>,
}

impl LockFile {
    /// Constructs and locks the file at the specified path. Blocks until the lock file is acquired.
    pub fn new(path: impl AsRef<Path>) -> anyhow::Result<Self> {
        // Ensure the directory exists
        let path = path.as_ref();
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }

        tracing::debug!("acquiring lockfile at '{}'", path.display());

        // Acquire the lock
        let mut lock = fslock::LockFile::open(path)?;

        // Block until the lock can be acquired
        lock.lock()?;

        Ok(LockFile {
            path: path.canonicalize()?,
            lock: Some(lock),
        })
    }

    /// Constructs and locks the file at the specified. Asynchronously waits until the file is
    /// locked.
    pub async fn new_async(path: impl AsRef<Path>) -> anyhow::Result<Self> {
        let path = path.as_ref().to_path_buf();
        tokio::task::spawn_blocking(move || Self::new(path)).await?
    }
}

impl Drop for LockFile {
    fn drop(&mut self) {
        drop(self.lock.take());

        // Ignore an error during deletion of the file. If another process acquired the lock this is
        // fine. Worst case an empty file remains.
        let _ = std::fs::remove_file(&self.path);
    }
}

#[cfg(test)]
mod test {
    use super::LockFile;
    use std::path::PathBuf;

    #[test]
    fn test_lock_file() {
        let path = PathBuf::from("test.lock");

        // Create the lock file, it should exist
        let lock_file = LockFile::new(&path).unwrap();
        assert!(path.exists());

        // Try opening the lock file again, this should fail
        let mut new_lock_file = fslock::LockFile::open(&path).unwrap();
        assert!(!new_lock_file.try_lock().unwrap());

        // Drop the lock file, the lock file should be gone
        drop(lock_file);
        assert!(!path.exists());
    }
}
