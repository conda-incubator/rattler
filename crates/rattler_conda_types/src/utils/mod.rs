pub(crate) mod serde;
pub(crate) mod path;
