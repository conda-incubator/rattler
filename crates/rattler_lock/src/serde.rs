use super::{CondaLock, LockMeta, LockedDependency, LockedDependencyKind};
use serde::de::Error;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::cmp::Ordering;
use std::path::PathBuf;

// Version 2: dependencies are now arrays instead of maps
// Version 3: pip has been renamed to pypi
const FILE_VERSION: u32 = 3;

/// A helper struct to deserialize the version field of the lock file and provide potential errors
/// in-line.
#[derive(Serialize)]
#[serde(transparent)]
struct Version(u32);

impl Default for Version {
    fn default() -> Self {
        Self(FILE_VERSION)
    }
}

impl<'de> Deserialize<'de> for Version {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let version = u32::deserialize(deserializer)?;

        if version > FILE_VERSION {
            let binary = if cfg!(test) {
                // When running tests, use a hardcoded name or an environment variable
                String::from(env!("CARGO_PKG_NAME"))
            } else {
                // When not testing, use the current executable's name, e.g. "pixi" for pixi.
                let binary_path =
                    std::env::current_exe().unwrap_or(PathBuf::from(env!("CARGO_PKG_NAME")));
                let binary_file_name = binary_path
                    .file_name()
                    .unwrap_or(env!("CARGO_PKG_NAME").as_ref());
                binary_file_name.to_string_lossy().into_owned()
            };
            return Err(D::Error::custom(format!(
                "found newer lockfile format version {}, but only up to including version {} is supported.\n\
                Please update your {} version.",
                version, FILE_VERSION, binary
            )));
        }

        Ok(Self(version))
    }
}

impl<'de> Deserialize<'de> for CondaLock {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[allow(dead_code)]
        #[derive(Deserialize)]
        struct Raw {
            version: Version,
            metadata: LockMeta,
            package: Vec<LockedDependency>,
        }

        let raw = Raw::deserialize(deserializer)?;
        Ok(Self {
            metadata: raw.metadata,
            package: raw.package,
        })
    }
}

impl Serialize for CondaLock {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        #[derive(Serialize)]
        struct Raw<'a> {
            version: Version,
            metadata: &'a LockMeta,
            package: Vec<&'a LockedDependency>,
        }

        // Sort all packages in alphabetical order. We choose to use alphabetic order instead of
        // topological because the alphabetic order will create smaller diffs when packages change
        // or are added.
        // See: https://github.com/conda/conda-lock/issues/491
        let mut sorted_deps = self.package.iter().collect::<Vec<_>>();
        sorted_deps.sort_by(|&a, &b| {
            a.name
                .cmp(&b.name)
                .then_with(|| a.platform.cmp(&b.platform))
                .then_with(|| a.version.cmp(&b.version))
                .then_with(|| match (&a.kind, &b.kind) {
                    (LockedDependencyKind::Conda(a), LockedDependencyKind::Conda(b)) => {
                        a.build.cmp(&b.build)
                    }
                    (LockedDependencyKind::Pypi(_), LockedDependencyKind::Pypi(_)) => {
                        Ordering::Equal
                    }
                    (LockedDependencyKind::Pypi(_), _) => Ordering::Less,
                    (_, LockedDependencyKind::Pypi(_)) => Ordering::Greater,
                })
        });

        let raw = Raw {
            version: Default::default(),
            metadata: &self.metadata,
            package: sorted_deps,
        };

        raw.serialize(serializer)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use std::path::Path;

    #[test]
    fn read_conda_lock() {
        let err = CondaLock::from_path(
            &Path::new(env!("CARGO_MANIFEST_DIR"))
                .join("../../test-data/conda-lock/forward-compatible-lock.yml"),
        )
        .unwrap_err();

        insta::assert_snapshot!(format!("{}", err), @"found newer lockfile format version 1000, but only up to including version 3 is supported.\nPlease update your rattler_lock version.");
    }
}
