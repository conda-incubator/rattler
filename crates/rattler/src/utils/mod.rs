use std::fmt::Write;

use url::Url;

pub use encoding::{AsyncEncoding, Encoding};

mod encoding;
mod lock_file;

pub use lock_file::LockFile;

#[cfg(test)]
pub(crate) mod simple_channel_server;

/// Convert a URL to a cache filename
pub fn url_to_cache_filename(url: &Url) -> String {
    // Start Rant:
    // This function mimics behavior from Mamba which itself mimics this behavior from Conda.
    // However, I find this function absolutely ridiculous, it contains all sort of weird edge
    // cases and returns a hash that could very easily collide with other files. And why? Why not
    // simply return a little more descriptive? Like a URL encoded string?
    // End Rant.
    let mut url_str = url.to_string();

    // Ensure there is a slash if the URL is empty or doesnt refer to json file
    if url_str.is_empty() || (!url_str.ends_with('/') && !url_str.ends_with(".json")) {
        url_str.push('/')
    }

    // Mimicking conda's (weird) behavior by special handling repodata.json
    let url_str = url_str.strip_suffix("/repodata.json").unwrap_or(&url_str);

    // Compute the MD5 hash of the resulting URL string
    let hash = extendhash::md5::compute_hash(url_str.as_bytes());

    // Convert the hash to an MD5 hash.
    let mut result = String::with_capacity(8);
    for x in &hash[0..4] {
        write!(result, "{:02x}", x).unwrap();
    }
    result
}

macro_rules! regex {
    ($re:literal $(,)?) => {{
        static RE: once_cell::sync::OnceCell<regex::Regex> = once_cell::sync::OnceCell::new();
        RE.get_or_init(|| regex::Regex::new($re).unwrap())
    }};
}

pub use regex;

#[cfg(test)]
mod test {
    use url::Url;

    use crate::utils::url_to_cache_filename;

    #[test]
    fn test_url_to_cache_filename() {
        assert_eq!(
            url_to_cache_filename(&Url::parse("http://test.com/1234/").unwrap()),
            "302f0a61"
        );
    }
}
