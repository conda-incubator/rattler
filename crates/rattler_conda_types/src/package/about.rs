use std::{path::Path, fs::File, io::Read, str::FromStr};

use crate::utils::serde::{LossyUrl, MultiLineString, VecSkipNone};
use serde::{Deserialize, Serialize};
use serde_with::{serde_as, skip_serializing_none, OneOrMany, Same};

use url::Url;

#[serde_as]
#[skip_serializing_none]
#[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
pub struct About {
    /// Description of the package
    #[serde_as(deserialize_as = "Option<MultiLineString>")]
    pub description: Option<String>,

    /// Short summary description
    #[serde_as(deserialize_as = "Option<MultiLineString>")]
    pub summary: Option<String>,

    /// Optionally, the license
    pub license: Option<String>,

    /// Optionally, the license family
    pub license_family: Option<String>,

    /// URL to the development page of the package
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    #[serde_as(
        deserialize_as = "VecSkipNone<OneOrMany<LossyUrl>>",
        serialize_as = "OneOrMany<Same>"
    )]
    pub dev_url: Vec<Url>,

    /// URL to the documentation of the package
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    #[serde_as(
        deserialize_as = "VecSkipNone<OneOrMany<LossyUrl>>",
        serialize_as = "OneOrMany<Same>"
    )]
    pub doc_url: Vec<Url>,

    /// URL to the homepage of the package
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    #[serde_as(
        deserialize_as = "VecSkipNone<OneOrMany<LossyUrl>>",
        serialize_as = "OneOrMany<Same>"
    )]
    pub home: Vec<Url>,

    /// URL to the latest source code of the package
    #[serde(default)]
    #[serde_as(deserialize_as = "LossyUrl")]
    pub source_url: Option<Url>,

    /// A list of channels that where used during the build
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub channels: Vec<String>,
}

impl About {
    /// Parses a `about.json` file from a reader.
    pub fn from_reader(mut reader: impl Read) -> Result<Self, std::io::Error> {
        let mut str = String::new();
        reader.read_to_string(&mut str)?;
        Self::from_str(&str)
    }

    /// Parses a `about.json` file from a file.
    pub fn from_path(path: &Path) -> Result<Self, std::io::Error> {
        Self::from_reader(File::open(path)?)
    }

    /// Reads the file from a package archive directory
    pub fn from_package_directory(path: &Path) -> Result<Self, std::io::Error> {
        Self::from_path(&path.join("info/about.json"))
    }
}

impl FromStr for About {
    type Err = std::io::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        serde_json::from_str(s).map_err(Into::into)
    }
}

#[cfg(test)]
mod test {
    use super::About;

    #[test]
    pub fn test_reconstruct_about_json() {
        let package_dir = tempfile::tempdir().unwrap();
        rattler_package_streaming::fs::extract(
            &crate::get_test_data_dir().join("conda-22.11.1-py38haa244fe_1.conda"),
            package_dir.path(),
        )
        .unwrap();

        insta::assert_yaml_snapshot!(About::from_package_directory(package_dir.path()).unwrap());
    }

    #[test]
    pub fn test_reconstruct_about_json_mamba() {
        let package_dir = tempfile::tempdir().unwrap();
        rattler_package_streaming::fs::extract(
            &crate::get_test_data_dir().join("mamba-1.0.0-py38hecfeebb_2.tar.bz2"),
            package_dir.path(),
        )
        .unwrap();

        let package_dir = package_dir.into_path();
        println!("{}", package_dir.display());

        insta::assert_yaml_snapshot!(About::from_package_directory(&package_dir).unwrap());
    }
}
