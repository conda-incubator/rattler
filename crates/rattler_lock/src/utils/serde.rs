use chrono::{DateTime, Utc};
use fxhash::FxBuildHasher;
use indexmap::IndexMap;
use pep508_rs::VersionOrUrl;
use rattler_conda_types::MatchSpec;
use rattler_conda_types::{NamelessMatchSpec, PackageName};
use serde::de::Error as _;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::de::DeserializeAsWrap;
use serde_with::ser::SerializeAsWrap;
use serde_with::DisplayFromStr;
use serde_with::{serde_as, DeserializeAs, SerializeAs};
use std::collections::HashSet;
use std::hash::{BuildHasher, Hash};
use std::marker::PhantomData;

pub(crate) struct Timestamp;

impl<'de> DeserializeAs<'de, chrono::DateTime<chrono::Utc>> for Timestamp {
    fn deserialize_as<D>(deserializer: D) -> Result<chrono::DateTime<chrono::Utc>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let timestamp = i64::deserialize(deserializer)?;

        // Convert from milliseconds to seconds
        let microseconds = if timestamp > 253_402_300_799 {
            timestamp * 1_000
        } else {
            timestamp * 1_000_000
        };

        // Convert the timestamp to a UTC timestamp
        Ok(chrono::DateTime::<chrono::Utc>::from_naive_utc_and_offset(
            chrono::NaiveDateTime::from_timestamp_micros(microseconds)
                .ok_or_else(|| D::Error::custom("got invalid timestamp, timestamp out of range"))?,
            chrono::Utc,
        ))
    }
}

impl SerializeAs<chrono::DateTime<chrono::Utc>> for Timestamp {
    fn serialize_as<S>(source: &DateTime<Utc>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        // Convert the date to a timestamp
        let timestamp: i64 = source.timestamp_millis();

        // Determine the precision of the timestamp.
        let timestamp = if timestamp % 1000 == 0 {
            timestamp / 1000
        } else {
            timestamp
        };

        // Serialize the timestamp
        timestamp.serialize(serializer)
    }
}

/// Used with serde_with to serialize a collection as a sorted collection.
#[derive(Default)]
pub(crate) struct Ordered<T>(PhantomData<T>);

impl<'de, T: Eq + Hash, S: BuildHasher + Default, TAs> DeserializeAs<'de, HashSet<T, S>>
    for Ordered<TAs>
where
    TAs: DeserializeAs<'de, T>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<HashSet<T, S>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let content =
            DeserializeAsWrap::<Vec<T>, Vec<TAs>>::deserialize(deserializer)?.into_inner();
        Ok(HashSet::from_iter(content))
    }
}

impl<T: Ord, HS, TAs: SerializeAs<T>> SerializeAs<HashSet<T, HS>> for Ordered<TAs> {
    fn serialize_as<S>(source: &HashSet<T, HS>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut elements = Vec::from_iter(source.iter());
        elements.sort();
        SerializeAsWrap::<Vec<&T>, Vec<&TAs>>::new(&elements).serialize(serializer)
    }
}

impl<'de, T: Ord, TAs> DeserializeAs<'de, Vec<T>> for Ordered<TAs>
where
    TAs: DeserializeAs<'de, T>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<T>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let mut content =
            DeserializeAsWrap::<Vec<T>, Vec<TAs>>::deserialize(deserializer)?.into_inner();
        content.sort();
        Ok(content)
    }
}

impl<T: Ord, TAs: SerializeAs<T>> SerializeAs<Vec<T>> for Ordered<TAs> {
    fn serialize_as<S>(source: &Vec<T>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut elements = Vec::from_iter(source.iter());
        elements.sort();
        SerializeAsWrap::<Vec<&T>, Vec<&TAs>>::new(&elements).serialize(serializer)
    }
}

pub(crate) struct MatchSpecMapOrVec;

impl<'de> DeserializeAs<'de, Vec<String>> for MatchSpecMapOrVec {
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<String>, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[serde_as]
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum MapOrVec {
            Vec(Vec<String>),
            Map(
                #[serde_as(as = "IndexMap<_, DisplayFromStr, FxBuildHasher>")]
                IndexMap<PackageName, NamelessMatchSpec, FxBuildHasher>,
            ),
        }

        Ok(match MapOrVec::deserialize(deserializer)? {
            MapOrVec::Vec(v) => v,
            MapOrVec::Map(m) => m
                .into_iter()
                .map(|(name, spec)| MatchSpec::from_nameless(spec, Some(name)).to_string())
                .collect(),
        })
    }
}

pub(crate) struct Pep440MapOrVec;

impl<'de> DeserializeAs<'de, Vec<String>> for Pep440MapOrVec {
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<String>, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[serde_as]
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum MapOrVec {
            Vec(Vec<String>),
            Map(
                #[serde_as(as = "IndexMap<_, DisplayFromStr, FxBuildHasher>")]
                IndexMap<String, pep440_rs::VersionSpecifiers, FxBuildHasher>,
            ),
        }

        Ok(match MapOrVec::deserialize(deserializer)? {
            MapOrVec::Vec(v) => v,
            MapOrVec::Map(m) => m
                .into_iter()
                .map(|(name, spec)| {
                    pep508_rs::Requirement {
                        name,
                        extras: None,
                        version_or_url: if spec.is_empty() {
                            None
                        } else {
                            Some(VersionOrUrl::VersionSpecifier(spec))
                        },
                        marker: None,
                    }
                    .to_string()
                })
                .collect(),
        })
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use serde::Deserialize;
    use serde_with::serde_as;

    #[test]
    fn test_parse_dependencies_as_map_or_vec() {
        #[serde_as]
        #[derive(Deserialize, Eq, PartialEq, Debug)]
        struct Data {
            #[serde_as(deserialize_as = "MatchSpecMapOrVec")]
            dependencies: Vec<String>,
        }

        let data_with_map: Data = serde_yaml::from_str("dependencies:\n  foo: \">3.12\"").unwrap();
        let data_with_vec: Data = serde_yaml::from_str("dependencies:\n- \"foo >3.12\"").unwrap();
        assert_eq!(data_with_map, data_with_vec);
    }
}
