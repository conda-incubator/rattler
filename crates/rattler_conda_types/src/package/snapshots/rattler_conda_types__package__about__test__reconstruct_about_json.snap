---
source: crates/rattler_conda_types/src/package/about.rs
expression: "About::from_package_directory(package_dir.path()).unwrap()"
---
description: "Conda is an open source package management system and environment management system for installing multiple versions of software packages and their dependencies and switching easily between them. It works on Linux, OS X and Windows, and was created for Python programs but can package and distribute any software.\n"
summary: "OS-agnostic, system-level binary package and environment manager."
license: BSD-3-Clause
dev_url: "https://github.com/conda/conda"
doc_url: "https://docs.conda.io/projects/conda/en/stable/"
home: "https://conda.io/"
channels:
  - "https://conda.anaconda.org/conda-forge"

